bitpacker/src/blocked_bitpacker.rs

```diff
@@ -0,0 +1,176 @@
+use crate::BitUnpacker;
+
+use super::{bitpacker::BitPacker, compute_num_bits};
+
+const BLOCK_SIZE: usize = 128;
+
+/// `BlockedBitpacker` compresses data in blocks of
+/// 128 elements, while keeping an index on it
+///
+#[derive(Debug, Clone)]
+pub struct BlockedBitpacker {
+    // bitpacked blocks
+    compressed_blocks: Vec<u8>,
+    // uncompressed data, collected until BLOCK_SIZE
+    cache: Vec<u64>,
+    offset_and_bits: Vec<BlockedBitpackerEntryMetaData>,
+}
+
+/// `BlockedBitpackerEntryMetaData` encodes the
+/// offset and bit_width into a u64 bit field
+///
+/// This saves some space, since 7byte is more
+/// than enough and also keeps the access fast
+/// because of alignment
+#[derive(Debug, Clone, Default)]
+struct BlockedBitpackerEntryMetaData {
+    encoded: u64,
+    base_value: u64,
+}
+
+impl BlockedBitpackerEntryMetaData {
+    fn new(offset: u64, num_bits: u8, base_value: u64) -> Self {
+        let encoded = offset | (num_bits as u64) << (64 - 8);
+        Self {
+            encoded,
+            base_value,
+        }
+    }
+    fn offset(&self) -> u64 {
+        (self.encoded << 8) >> 8
+    }
+    fn num_bits(&self) -> u8 {
+        (self.encoded >> 56) as u8
+    }
+    fn base_value(&self) -> u64 {
+        self.base_value
+    }
+}
+
+#[test]
+fn metadata_test() {
+    let meta = BlockedBitpackerEntryMetaData::new(50000, 6, 40000);
+    assert_eq!(meta.offset(), 50000);
+    assert_eq!(meta.num_bits(), 6);
+}
+
+impl BlockedBitpacker {
+    pub fn new() -> Self {
+        let mut compressed_blocks = vec![];
+        compressed_blocks.resize(8, 0);
+        Self {
+            compressed_blocks,
+            cache: vec![],
+            offset_and_bits: vec![],
+        }
+    }
+
+    /// The memory used (inclusive childs)
+    pub fn mem_usage(&self) -> usize {
+        std::mem::size_of::<BlockedBitpacker>()
+            + self.compressed_blocks.capacity()
+            + self.offset_and_bits.capacity()
+                * std::mem::size_of_val(&self.offset_and_bits.get(0).cloned().unwrap_or_default())
+            + self.cache.capacity()
+                * std::mem::size_of_val(&self.cache.get(0).cloned().unwrap_or_default())
+    }
+
+    pub fn add(&mut self, val: u64) {
+        self.cache.push(val);
+        if self.cache.len() == BLOCK_SIZE as usize {
+            self.flush();
+        }
+    }
+
+    pub fn flush(&mut self) {
+        if self.cache.is_empty() {
```

Comment:
nitpick...

nowadays I like to avoid the unwrap using the following pattern.

let base_value = if let Some(min_value) = self.cache.iter().min() {
   min_value
} else {
   return;
};

Reply:
I can do that, but there is still a unwrap() in line 96